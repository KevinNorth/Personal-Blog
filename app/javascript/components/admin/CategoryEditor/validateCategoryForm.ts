export type Validation = {
  isValid: false;
  invalidReason: string;
} | {
  isValid: true;
  invalidReason: undefined;
}

export interface ValidationResults {
  markdown: Validation;
  name: Validation;
  order: Validation;
  published: Validation;
  slug: Validation;
  subtitle: Validation;
  summary: Validation;
  title: Validation;
}

function validateCategoryForm({
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  markdown,
  name,
  order,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  published,
  slug,
  subtitle,
  summary,
  title,
  usedSlugs,
  usedOrders
}: {
  markdown: string,
  name: string,
  order: string,
  published: boolean,
  slug: string,
  subtitle: string,
  summary: string,
  title: string,
  usedSlugs: string[],
  usedOrders: number[]
}): ValidationResults {
  const validationResults: ValidationResults =
    ['markdown', 'name', 'order', 'published', 'slug', 'subtitle', 'summary', 'title']
      .reduce(
        (allResults: Partial<ValidationResults>, key: string) => ({
          ...allResults,
          [key]: { isValid: true }
        }),
        {}
      ) as ValidationResults;

  if (title === '') {
    validationResults.title = {
      isValid: false,
      invalidReason: 'Title must not be blank.'
    };
  }

  if (subtitle === '') {
    validationResults.subtitle = {
      isValid: false,
      invalidReason: 'Subtitle must not be blank.'
    };
  }

  if (summary === '') {
    validationResults.summary = {
      isValid: false,
      invalidReason: 'Summary must not be blank.'
    };
  }

  if (name === '') {
    validationResults.name = {
      isValid: false,
      invalidReason: 'Name must not be blank.'
    };
  }

  if (order === '') {
    validationResults.order = {
      isValid: false,
      invalidReason: 'Order must not be blank.'
    };
  } else if (isNaN(+order)) {
    validationResults.order = {
      isValid: false,
      invalidReason: 'Order must be a number.'
    };
  } else if (usedOrders.includes(Number(order))) {
    validationResults.order = {
      isValid: false,
      invalidReason: "Cannot match another category's order under the same parent."
    };
  }

  if (slug === '') {
    validationResults.slug = {
      isValid: false,
      invalidReason: 'Slug must not be blank.'
    };
  } else if (usedSlugs.includes(slug)) {
    validationResults.slug = {
      isValid: false,
      invalidReason: 'Slug is already used by another category.'
    };
  }
  
  return validationResults;
}

export default validateCategoryForm;
