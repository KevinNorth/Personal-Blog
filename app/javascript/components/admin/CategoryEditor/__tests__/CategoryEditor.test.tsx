/**
 * @jest-environment jsdom
 */
import React from 'react';
import snapshotRenderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import {
  act,
  render as interactableRenderer,
  RenderResult,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mockCategoriesAndPosts } from '../../../../__tests__/fixtures/allCategoriesAndPosts';
import { allCategoriesAndPostsQuery } from '../../../../graphql/queries/allCategoriesAndPosts';
import CategoryEditor, { CategoryEditorProps } from '../CategoryEditor';

function TestCategoryEditor(
  props: Partial<CategoryEditorProps> & { key?: string | number }
): React.ReactElement {
  return (
    <CategoryEditor
      id={props.id || 'test-category-editor'}
      loading={props.loading === undefined ? false : props.loading}
      markdown={props.markdown || 'Test markdown'}
      name={props.name || 'Test name'}
      order={props.order || '1'}
      parentId={props.parentId || null}
      published={props.published === undefined ? true : props.published}
      slug={props.slug || 'test-slug'}
      subtitle={props.subtitle || 'Test subtitle'}
      summary={props.summary || 'Test summary'}
      title={props.title || 'Test title'}
      onMarkdownChange={props.onMarkdownChange || jest.fn()}
      onNameChange={props.onNameChange || jest.fn()}
      onOrderChange={props.onOrderChange || jest.fn()}
      onParentIdChange={props.onParentIdChange || jest.fn()}
      onPublishedChange={props.onPublishedChange || jest.fn()}
      onSlugChange={props.onSlugChange || jest.fn()}
      onSubtitleChange={props.onSubtitleChange || jest.fn()}
      onSummaryChange={props.onSummaryChange || jest.fn()}
      onTitleChange={props.onTitleChange || jest.fn()}
      key={props.key || undefined}
    >
      {props.children || <span>Test children</span>}
    </CategoryEditor>
  );
}

function testTextInput(input: keyof CategoryEditorProps, label?: string) {
  const capitalizedInput = input.charAt(0).toUpperCase() + input.slice(1);

  it(`calls on${capitalizedInput}Change`, async () => {
    const user = userEvent.setup();
    const onChangeSpy = jest.fn();
    const initialValue = '1';
    const typedChar = '2';

    const props = {
      [input]: initialValue,
      [`on${capitalizedInput}Change`]: onChangeSpy,
    };

    let wrapper: RenderResult;
    act(() => {
      wrapper = interactableRenderer(
        <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
          <TestCategoryEditor {...props} />
        </MockedProvider>
      );
    });

    const escapedLabelForRegExp = (label || capitalizedInput).replace(
      /[/\-\\^$*+?.()|[\]{}]/g,
      '\\$&'
    );
    // We are in complete control of the values passed to the regexp here,
    // and we escape characters with regex meaning as well.
    // eslint-disable-next-line security/detect-non-literal-regexp
    const labelRegExp = new RegExp(`^${escapedLabelForRegExp}$`);

    await waitFor(() => {
      // Some form inputs are only available after GraphQL queries resolve
      expect(wrapper.findByLabelText(labelRegExp)).toBeTruthy();
    });
    const textbox = screen.getByLabelText(labelRegExp);
    await user.type(textbox, typedChar);

    expect(onChangeSpy.mock.lastCall?.[0]).toEqual(
      `${initialValue}${typedChar}`
    );
  });
}

const getAllCategoriesAndPostsMock = {
  request: {
    query: allCategoriesAndPostsQuery,
    variables: {
      includeUnpublished: true,
    },
  },
  result: {
    data: {
      categories: mockCategoriesAndPosts,
    },
  },
};

describe('CategoryEditor', () => {
  describe('when loading', () => {
    it('renders as expected', () => {
      let wrapper;

      act(() => {
        wrapper = snapshotRenderer.create(
          <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
            <TestCategoryEditor loading={true} />
          </MockedProvider>
        );
      });

      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('when given a valid category form', () => {
    describe('when the category is a top-level category', () => {
      it('renders as expected', () => {
        let wrapper;

        act(() => {
          wrapper = snapshotRenderer.create(
            <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
              <TestCategoryEditor parentId={null} loading={false} />
            </MockedProvider>
          );
        });

        expect(wrapper.toJSON()).toMatchSnapshot();
      });
    });

    describe('when the category is a child category', () => {
      it('renders as expected', () => {
        let wrapper;

        act(() => {
          wrapper = snapshotRenderer.create(
            <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
              <TestCategoryEditor parentId="1" loading={false} />
            </MockedProvider>
          );
        });

        expect(wrapper.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('when given an invalid category form', () => {
    it('renders as expected', () => {
      let wrapper;

      act(() => {
        wrapper = snapshotRenderer.create(
          <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
            <TestCategoryEditor order="invalid" loading={false} />
          </MockedProvider>
        );
      });

      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('user interactions', () => {
    describe('when the user enters markdown', () => {
      it('calls onMarkdownChange', async () => {
        const user = userEvent.setup();
        const onMarkdownChangeSpy = jest.fn();
        const initialMarkdown = 'a';
        const typedChar = 'b';

        let wrapper;
        act(() => {
          wrapper = interactableRenderer(
            <MockedProvider mocks={[getAllCategoriesAndPostsMock]}>
              <TestCategoryEditor
                markdown={initialMarkdown}
                onMarkdownChange={onMarkdownChangeSpy}
              />
            </MockedProvider>
          );
        });

        expect(
          (await wrapper.findAllByDisplayValue(initialMarkdown)).length
        ).toBeGreaterThan(0);
        const markdownInput = screen.getAllByDisplayValue(initialMarkdown)[0];
        await user.type(markdownInput, typedChar);

        expect(onMarkdownChangeSpy.mock.lastCall?.[0]).toEqual(
          `${initialMarkdown}${typedChar}`
        );
      });
    });

    describe('when the user enters a name', () => {
      testTextInput('name', 'Name in Navbar');
    });

    describe('when the user enters an order', () => {
      testTextInput('order', 'Order in Navbar');
    });

    describe('when the user selects a parent category', () => {
      describe('when the user selects a non-null parent category', () => {
        test.todo('calls onParentIdChange');
      });

      describe('when the user selects "No parent" for parent category', () => {
        test.todo('calls onParentIdChange with null');
      });
    });

    describe('when the user clicks on a published button', () => {
      describe('when the user clicks "Published', () => {
        test.todo('calls onPublishedChange(true)');
      });

      describe('when the user clicks "Unpublished', () => {
        test.todo('calls onPublishedChange(false)');
      });
    });

    describe('when the user enters a slug', () => {
      testTextInput('slug');
    });

    describe('when the user enters a subtitle', () => {
      testTextInput('subtitle');
    });

    describe('when the user enters a summary', () => {
      testTextInput('summary');
    });

    describe('when the user enters a title', () => {
      testTextInput('title');
    });
  });
});
