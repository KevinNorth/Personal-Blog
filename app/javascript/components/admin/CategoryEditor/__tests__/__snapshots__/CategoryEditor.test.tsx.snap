// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryEditor when given a valid category form when the category is a child category renders as expected 1`] = `
<div
  className="category-editor container-fluid"
>
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-title"
        >
          <label
            className="form-label"
            htmlFor="category-title"
            size="lg"
          >
            Title
          </label>
          <input
            className="form-control form-control-lg is-valid"
            id="category-title"
            onChange={[Function]}
            type="text"
            value="Test title"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-subtitle"
        >
          <label
            className="form-label"
            htmlFor="category-subtitle"
          >
            Subtitle
          </label>
          <input
            className="form-control is-valid"
            id="category-subtitle"
            onChange={[Function]}
            type="text"
            value="Test subtitle"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-summary"
        >
          <label
            className="form-label"
            htmlFor="category-summary"
          >
            Summary
          </label>
          <input
            className="form-control is-valid"
            id="category-summary"
            onChange={[Function]}
            type="text"
            value="Test summary"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="select-group category-parent"
        >
          <label
            className="form-label"
            htmlFor="category-parent"
          >
            Category
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-6"
      >
        <div
          className="button-group category-published"
        >
          <div
            className="btn-group"
            role="group"
          >
            <input
              autoComplete="off"
              checked={false}
              className="btn-check"
              disabled={false}
              id="category-unpublished-button"
              name="unpublished"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-unpublished-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Unpublished
            </label>
            <input
              autoComplete="off"
              checked={true}
              className="btn-check"
              disabled={false}
              id="category-published-button"
              name="published"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-published-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Published
            </label>
          </div>
        </div>
      </div>
      <div
        className="col-6"
      >
        <div
          className="text-group category-slug"
        >
          <label
            className="form-label"
            htmlFor="category-slug"
          >
            Slug
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-name"
        >
          <label
            className="form-label"
            htmlFor="category-name"
          >
            Name in Navbar
          </label>
          <input
            className="form-control is-valid"
            id="category-name"
            onChange={[Function]}
            type="text"
            value="Test name"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="number-group category-order"
        >
          <label
            className="form-label"
            htmlFor="category-order"
          >
            Order in Navbar
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="editor-group category-body"
        >
          <ul
            className="editor-tabs nav nav-tabs"
            onKeyDown={[Function]}
            role="tablist"
          >
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-split"
                aria-selected={true}
                className="nav-link active"
                data-rr-ui-event-key="split"
                disabled={false}
                id="core-markdown-editor-tab-split"
                onClick={[Function]}
                role="tab"
                type="button"
              >
                Split Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-editor"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="editor"
                disabled={false}
                id="core-markdown-editor-tab-editor"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-preview"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="preview"
                disabled={false}
                id="core-markdown-editor-tab-preview"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Preview
              </button>
            </li>
          </ul>
          <div
            className="tab-content"
          >
            <div
              aria-labelledby="core-markdown-editor-tab-split"
              className="tab-pane active"
              id="core-markdown-editor-tabpane-split"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-6"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                  <div
                    className="editor-preview-wrapper col-6"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-editor"
              className="tab-pane"
              id="core-markdown-editor-tabpane-editor"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-12"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-preview"
              className="tab-pane"
              id="core-markdown-editor-tabpane-preview"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-preview-wrapper col-12"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <footer
        className="col-12"
      >
        <span>
          Test children
        </span>
      </footer>
    </div>
  </form>
</div>
`;

exports[`CategoryEditor when given a valid category form when the category is a top-level category renders as expected 1`] = `
<div
  className="category-editor container-fluid"
>
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-title"
        >
          <label
            className="form-label"
            htmlFor="category-title"
            size="lg"
          >
            Title
          </label>
          <input
            className="form-control form-control-lg is-valid"
            id="category-title"
            onChange={[Function]}
            type="text"
            value="Test title"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-subtitle"
        >
          <label
            className="form-label"
            htmlFor="category-subtitle"
          >
            Subtitle
          </label>
          <input
            className="form-control is-valid"
            id="category-subtitle"
            onChange={[Function]}
            type="text"
            value="Test subtitle"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-summary"
        >
          <label
            className="form-label"
            htmlFor="category-summary"
          >
            Summary
          </label>
          <input
            className="form-control is-valid"
            id="category-summary"
            onChange={[Function]}
            type="text"
            value="Test summary"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="select-group category-parent"
        >
          <label
            className="form-label"
            htmlFor="category-parent"
          >
            Category
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-6"
      >
        <div
          className="button-group category-published"
        >
          <div
            className="btn-group"
            role="group"
          >
            <input
              autoComplete="off"
              checked={false}
              className="btn-check"
              disabled={false}
              id="category-unpublished-button"
              name="unpublished"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-unpublished-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Unpublished
            </label>
            <input
              autoComplete="off"
              checked={true}
              className="btn-check"
              disabled={false}
              id="category-published-button"
              name="published"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-published-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Published
            </label>
          </div>
        </div>
      </div>
      <div
        className="col-6"
      >
        <div
          className="text-group category-slug"
        >
          <label
            className="form-label"
            htmlFor="category-slug"
          >
            Slug
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-name"
        >
          <label
            className="form-label"
            htmlFor="category-name"
          >
            Name in Navbar
          </label>
          <input
            className="form-control is-valid"
            id="category-name"
            onChange={[Function]}
            type="text"
            value="Test name"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="number-group category-order"
        >
          <label
            className="form-label"
            htmlFor="category-order"
          >
            Order in Navbar
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="editor-group category-body"
        >
          <ul
            className="editor-tabs nav nav-tabs"
            onKeyDown={[Function]}
            role="tablist"
          >
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-split"
                aria-selected={true}
                className="nav-link active"
                data-rr-ui-event-key="split"
                disabled={false}
                id="core-markdown-editor-tab-split"
                onClick={[Function]}
                role="tab"
                type="button"
              >
                Split Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-editor"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="editor"
                disabled={false}
                id="core-markdown-editor-tab-editor"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-preview"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="preview"
                disabled={false}
                id="core-markdown-editor-tab-preview"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Preview
              </button>
            </li>
          </ul>
          <div
            className="tab-content"
          >
            <div
              aria-labelledby="core-markdown-editor-tab-split"
              className="tab-pane active"
              id="core-markdown-editor-tabpane-split"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-6"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                  <div
                    className="editor-preview-wrapper col-6"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-editor"
              className="tab-pane"
              id="core-markdown-editor-tabpane-editor"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-12"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-preview"
              className="tab-pane"
              id="core-markdown-editor-tabpane-preview"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-preview-wrapper col-12"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <footer
        className="col-12"
      >
        <span>
          Test children
        </span>
      </footer>
    </div>
  </form>
</div>
`;

exports[`CategoryEditor when given an invalid category form renders as expected 1`] = `
<div
  className="category-editor container-fluid"
>
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-title"
        >
          <label
            className="form-label"
            htmlFor="category-title"
            size="lg"
          >
            Title
          </label>
          <input
            className="form-control form-control-lg is-valid"
            id="category-title"
            onChange={[Function]}
            type="text"
            value="Test title"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-subtitle"
        >
          <label
            className="form-label"
            htmlFor="category-subtitle"
          >
            Subtitle
          </label>
          <input
            className="form-control is-valid"
            id="category-subtitle"
            onChange={[Function]}
            type="text"
            value="Test subtitle"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-summary"
        >
          <label
            className="form-label"
            htmlFor="category-summary"
          >
            Summary
          </label>
          <input
            className="form-control is-valid"
            id="category-summary"
            onChange={[Function]}
            type="text"
            value="Test summary"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="select-group category-parent"
        >
          <label
            className="form-label"
            htmlFor="category-parent"
          >
            Category
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-6"
      >
        <div
          className="button-group category-published"
        >
          <div
            className="btn-group"
            role="group"
          >
            <input
              autoComplete="off"
              checked={false}
              className="btn-check"
              disabled={false}
              id="category-unpublished-button"
              name="unpublished"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-unpublished-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Unpublished
            </label>
            <input
              autoComplete="off"
              checked={true}
              className="btn-check"
              disabled={false}
              id="category-published-button"
              name="published"
              onChange={[Function]}
              type="radio"
              value="false"
            />
            <label
              className="btn btn-primary"
              disabled={false}
              htmlFor="category-published-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              Published
            </label>
          </div>
        </div>
      </div>
      <div
        className="col-6"
      >
        <div
          className="text-group category-slug"
        >
          <label
            className="form-label"
            htmlFor="category-slug"
          >
            Slug
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="text-group category-name"
        >
          <label
            className="form-label"
            htmlFor="category-name"
          >
            Name in Navbar
          </label>
          <input
            className="form-control is-valid"
            id="category-name"
            onChange={[Function]}
            type="text"
            value="Test name"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="number-group category-order"
        >
          <label
            className="form-label"
            htmlFor="category-order"
          >
            Order in Navbar
          </label>
          <span
            className="w-100 placeholder-glow"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <div
          className="editor-group category-body"
        >
          <ul
            className="editor-tabs nav nav-tabs"
            onKeyDown={[Function]}
            role="tablist"
          >
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-split"
                aria-selected={true}
                className="nav-link active"
                data-rr-ui-event-key="split"
                disabled={false}
                id="core-markdown-editor-tab-split"
                onClick={[Function]}
                role="tab"
                type="button"
              >
                Split Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-editor"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="editor"
                disabled={false}
                id="core-markdown-editor-tab-editor"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Editor
              </button>
            </li>
            <li
              className="nav-item"
              role="presentation"
            >
              <button
                aria-controls="core-markdown-editor-tabpane-preview"
                aria-selected={false}
                className="nav-link"
                data-rr-ui-event-key="preview"
                disabled={false}
                id="core-markdown-editor-tab-preview"
                onClick={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                Preview
              </button>
            </li>
          </ul>
          <div
            className="tab-content"
          >
            <div
              aria-labelledby="core-markdown-editor-tab-split"
              className="tab-pane active"
              id="core-markdown-editor-tabpane-split"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-6"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                  <div
                    className="editor-preview-wrapper col-6"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-editor"
              className="tab-pane"
              id="core-markdown-editor-tabpane-editor"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-input col-12"
                  >
                    <textarea
                      className="form-control"
                      id="category-body"
                      onChange={[Function]}
                      value="Test markdown"
                    />
                    <button
                      className="editor-syntax-help-button btn btn-primary"
                      disabled={false}
                      onBlur={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      Syntax Help
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="core-markdown-editor-tab-preview"
              className="tab-pane"
              id="core-markdown-editor-tabpane-preview"
              role="tabpanel"
            >
              <div
                className="category-markdown-editor container-fluid"
              >
                <div
                  className="row"
                >
                  <div
                    className="editor-preview-wrapper col-12"
                  >
                    <div
                      className="editor-preview"
                    >
                      <div
                        className="editor-preview-content"
                      >
                        <div>
                          Test markdown
                        </div>
                        <div>
                          Props: 
                          {"components":{},"remarkPlugins":[[null,{"singleTilde":false}]],"rehypePlugins":[null,null,null]}
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <footer
        className="col-12"
      >
        <span>
          Test children
        </span>
      </footer>
    </div>
  </form>
</div>
`;

exports[`CategoryEditor when loading renders as expected 1`] = `
<div
  className="container-fluid"
>
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <span
          className="w-100 placeholder-glow placeholder-lg"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <span
          className="w-100 placeholder-glow"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <span
          className="w-100 placeholder-glow"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-4"
      >
        <span
          className="w-100 placeholder-glow"
        />
      </div>
      <div
        className="col-8"
      >
        <span
          className="w-100 placeholder-glow"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <span
          className="w-100 placeholder-glow"
        />
      </div>
    </div>
  </form>
</div>
`;
